src/toxcore/dht/server/ping.rs

```diff
@@ -0,0 +1,236 @@
+pub struct Ping {
+    last_time_send_ping: Instant,
+    nodes_to_send_ping: RwLock<Bucket>,
+}
+
+impl Ping {
+    /// new Ping object
+    pub fn new() -> Self {
+        Ping {
+            last_time_send_ping: Instant::now(),
+            nodes_to_send_ping: RwLock::new(Bucket::new(None)),
+        }
+    }
+
+    fn is_bad_node_timed_out(node: &DhtNode, server: &Server) -> bool {
+        node.last_resp_time.elapsed() > Duration::from_secs(server.config.bad_node_timeout)
+    }
+
+    fn is_friend(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter().any(|friend| friend.pk == node.pk)
+    }
+
+    fn is_in_close_list(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter()
+            .map(|friend| friend.close_nodes.nodes.iter().any(|peer| peer.pk == node.pk))
+            .any(|result_bool| result_bool)
+    }
+
+    fn is_in_ping_list(node: &PackedNode, ping_list: &Bucket) -> bool {
+        ping_list.nodes.iter().any(|peer| peer.pk == node.pk)
+    }
+
+    fn is_iterate_time(&self, iterate_interval: Duration) -> bool {
+        self.last_time_send_ping.elapsed() >= iterate_interval
+    }
+
+    /// try to add node to list to send PingRequest
+    /// return true if node is added, false otherwise
+    pub fn try_add(&mut self, server: &Server, node: &PackedNode) -> bool {
+        // if node already exists in close list and not timed out, then don't send PingRequest
+        match server.close_nodes.read().find_node(&node.pk) {
+            Some(ref node_in_close_list) if !Ping::is_bad_node_timed_out(node_in_close_list, server) => return false,
+            _ => {},
+        };
+
+        // if node is not addable to close list, dont't send PingRequest
+        if !server.close_nodes.read().can_add(node) {
+            return false
+        }
+
+        // If node is friend and don't exist in friend's close list then send PingRequest
+        if Ping::is_friend(node, server) && !Ping::is_in_close_list(node, server) {
+            server.send_ping_req(node);
+            return false
+        }
+
+        let mut nodes_to_send_ping = self.nodes_to_send_ping.write();
+        // if node already exists in ping list, then don't add
+        if Ping::is_in_ping_list(node, nodes_to_send_ping.deref()) {
+            return false
+        }
+
+        // PingRequest is sent only for maximum 8 nodes in Bucket
+        nodes_to_send_ping.try_add(&server.pk, node)
+    }
+
+    /// send PingRequest to all nodes in list
+    pub fn iterate_ping_list(&mut self, server: &Server, iterate_interval: Duration) -> IoFuture<()> {
+        if !self.is_iterate_time(iterate_interval) {
+            return Box::new(future::ok(()))
+        }
+
+        let mut now = Instant::now();
+        let mut nodes_to_send_ping = Bucket::new(None);
+
+        mem::swap(&mut self.last_time_send_ping, &mut now);
+        mem::swap(self.nodes_to_send_ping.write().deref_mut(), &mut nodes_to_send_ping);
```

Comment:
Perhaps [mem::replace](https://doc.rust-lang.org/beta/std/mem/fn.replace.html) will be a bit nicer here.

Reply:
I want to use entries in current Bucket and then clear the Bucket.

Reply:
Right, and you can do it with replace:
```rust
let nodes_to_send_ping = mem::replace(self.nodes_to_send_ping.write().deref_mut(), Bucket::new(None))
```